# hotel cancellation prediction


importing basic libraries for spliting data in training and test set


In [1]:
#importing libraries
import pandas as pd 
import matplotlib.pyplot as plt

using pandas reading or importing dataset from my pc

In [2]:
data = pd.read_csv("C:\\Users\\fudan kumar\\Downloads\\hotel_bookings.csv.zip")

In [3]:
data .shape

(119390, 32)


i'm folding data for checking null value.

after folding or spliting in rowise ,
we can see null value of each columns.

you can also use heatmap diagram of seaborn but 
for less amount of column present in this data, 

so i'm using simple technique.



In [4]:
a = data.iloc[:,:10]
b = data.iloc[:,10:20]
c = data.iloc[:,20:]

In [5]:
a.isnull().sum()

hotel                        0
is_canceled                  0
lead_time                    0
arrival_date_year            0
arrival_date_month           0
arrival_date_week_number     0
arrival_date_day_of_month    0
stays_in_weekend_nights      0
stays_in_week_nights         0
adults                       0
dtype: int64

In [6]:
b.isnull().sum()

children                            4
babies                              0
meal                                0
country                           488
market_segment                      0
distribution_channel                0
is_repeated_guest                   0
previous_cancellations              0
previous_bookings_not_canceled      0
reserved_room_type                  0
dtype: int64

In [7]:
c.isnull().sum()

assigned_room_type                  0
booking_changes                     0
deposit_type                        0
agent                           16340
company                        112593
days_in_waiting_list                0
customer_type                       0
adr                                 0
required_car_parking_spaces         0
total_of_special_requests           0
reservation_status                  0
reservation_status_date             0
dtype: int64

we got null value as

children      4

country      488  

agent        16340

company      112593


In columns children,country and agent have less no of null value,

so i'm filling those value by mode of that columns

for filling you can also use other method 

like backward fill,and forward fill.


but in company columns have large amount of null value,

so i'm droping this column


In [8]:
data['children']= data['children'].fillna(data['children'].mode()[0])

In [9]:
data['country']= data['country'].fillna(data['country'].mode()[0])

In [10]:
data['agent']= data['agent'].fillna(data['agent'].mode()[0])

In [11]:
data.drop(['company'],inplace = True,axis =1)






after filling and droping columns
we should check null value is removed or not





In [12]:
data['agent'].isnull().sum()

0

spliting data into feature set and output/labels set


In [13]:
# deviding data into in the form feature and labels

#labels
y = data['is_canceled']

#feature
x1 = data.iloc[:,2:]
x2 = data['hotel']
x = pd.concat([x2,x1],axis = 1)

In our data have lots of catogorical value,

but we can't apply machine learning algorithm on string or catogorical dataset

so  i'm converting that data into numerical data set 

for converting catogorical data into numerical

you can use OneHotEncoder or labelencoder.

but i'm using one appication of pandas libraries for encoding.



i will recommend you to use OneHotEncoder.

In [14]:
# converting catogorical value
x1 = pd.get_dummies(x)
#deleting duplicate columns


for applying machine learning algorithm,

i'm spliting that data into trainig and test set.


i'm giving test_size = .25 that means 25% of data splited into testset.



In [15]:
# for applying algorithm spliting data into training and test set
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x1,y,random_state = 0,test_size = .25)

# using   DecisionTreeClassifier

In [16]:
from sklearn.tree import DecisionTreeClassifier
tree = DecisionTreeClassifier(random_state = 40,max_depth = 4)


In [17]:
tree.fit(x_train,y_train)
tree.score(x_test,y_test)


1.0

DecisionTreeClassifier is overfitting on this dataset.


after tuning of parameter we can get good model.


but i'm not tuning parameter in this model.


i'm going to use other machine learning algorithm.



In [18]:
from sklearn.neighbors import KNeighborsClassifier
clf = KNeighborsClassifier(n_neighbors = 3)

in this module i'm using n_neighbors= 3,

you can can use 3,4 and 5.
more than 5 you can't use becouse model becomes very complex.

In [19]:
clf.fit(x_train,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=3, p=2,
                     weights='uniform')

In [20]:
clf.score(x_test,y_test)

0.8010922004824443

In [21]:
clf.score(x_train,y_train)

0.9008956690714972

this is in  underfitting but you can get best score after tuning of parameter
you can also use hyperparameter.




...................................................... 